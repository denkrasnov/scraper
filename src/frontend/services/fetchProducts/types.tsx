import { ProductName, Article } from "../../../backend/scrapers/types";

export enum ActionTypes {
  FETCH = "FETCH",
  FETCH_INIT = "FETCH_INIT",
  FETCH_ERROR = "FETCH_ERROR",
  FETCH_SUCCESS = "FETCH_SUCCESS"
}

export interface State {
  isLoading: boolean;
  isError: string;
  query: string;
  products: Products | null;
}

export interface Fetch {
  type: ActionTypes.FETCH;
  payload: string;
}

interface ActionSuccess {
  type: ActionTypes.FETCH_SUCCESS;
  payload: Products;
}

interface ActionInit {
  type: ActionTypes.FETCH_INIT;
}

interface ActionError {
  type: ActionTypes.FETCH_ERROR;
  payload: any;
}

export type Action = ActionSuccess | ActionInit | ActionError | Fetch;

export interface Products {
  name: ProductName;
  items: Article[];
}

export interface Response {
  data: Products;
}
