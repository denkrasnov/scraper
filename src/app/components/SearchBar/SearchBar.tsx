import React, { FC } from "react";

import Box from "../../../components/Box";
import Button from "../../../components/Buttons";
import Input from "../../../components/Input";

const SearchBar: FC = () => {
  return (
    <>
      <Input placeholder="Search" />
      <Box padding="0 10px">
        <Button>search</Button>
      </Box>
    </>
  );
};

export default SearchBar;
