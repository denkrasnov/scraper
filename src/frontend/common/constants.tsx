import { Channel } from "../../backend/scrapers/types";

export const channelLogo: Record<string, string> = {
  [Channel.JurnalTV]:
    "https://live-tv-channels.org/pt-data/uploads/logo/md-jurnal-tv.jpg",
  [Channel.NTV]:
    "https://yt3.ggpht.com/ytc/AAUvwnjTxNZwMTMTlj18m75pswpomJFvJEHuargB5gWA=s900-c-k-c0x00ffffff-no-rj",
  [Channel.TV8]: "https://tv8.md/wp-content/uploads/2020/07/logo-tv8-new-2.png"
};
