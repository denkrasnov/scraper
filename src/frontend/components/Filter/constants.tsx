// TODO: created shared types with /backend
import { ProductName } from "../../../backend/scrapers/types";

export const filterName = {
  [ProductName.MD]: "Canal"
};
