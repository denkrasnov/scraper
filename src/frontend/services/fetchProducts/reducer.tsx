import { Action, State, ActionTypes } from "./types";

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case ActionTypes.FETCH:
      return {
        ...state,
        query: action.payload,
        isError: ""
      };
    case ActionTypes.FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: ""
      };
    case ActionTypes.FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        products: action.payload
      };
    case ActionTypes.FETCH_ERROR:
      return {
        ...state,
        isLoading: false,
        isError: action.payload
      };

    default:
      throw new Error();
  }
};

export default reducer;
