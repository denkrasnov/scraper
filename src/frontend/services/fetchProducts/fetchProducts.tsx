import { useEffect, useReducer, Dispatch } from "react";
import axios from "axios";

import { INITIAL_STATE } from "./constants";
import reducer from "./reducer";
import { Action, State, ActionTypes, Response } from "./types";

const fetchProducts = (): [State, Dispatch<Action>] => {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const { query } = state;

  useEffect(() => {
    let didCancel = false;

    const fetchData = async () => {
      dispatch({ type: ActionTypes.FETCH_INIT });
      try {
        const result: Response = await axios.get(`/search/${query}`);

        if (!didCancel) {
          dispatch({
            type: ActionTypes.FETCH_SUCCESS,
            payload: result.data.products
          });
        }
      } catch (error) {
        if (!didCancel) {
          dispatch({ type: ActionTypes.FETCH_ERROR, payload: error });
        }
      }
    };

    if (query) {
      fetchData();
    }

    return () => {
      didCancel = true;
    };
  }, [query]);

  return [state, dispatch];
};

export default fetchProducts;
