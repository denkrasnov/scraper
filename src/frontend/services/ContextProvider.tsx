import React, { FC, useContext, createContext } from "react";

import { INITIAL_STATE } from "./fetchProducts/constants";
import fetchProducts from "./fetchProducts";
import { ContextProviderProps, CreateContext } from "./types";

export const StateContext = createContext<CreateContext>([
  INITIAL_STATE,
  () => {}
]);

const ContextProvider: FC<ContextProviderProps> = ({ children }) => (
  <StateContext.Provider value={fetchProducts()}>
    {children}
  </StateContext.Provider>
);

export const useFullContext = () => useContext(StateContext);

export default ContextProvider;
