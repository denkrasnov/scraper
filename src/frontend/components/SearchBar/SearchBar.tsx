import React, { FC } from "react";
import { Form, Field } from "react-final-form";

import Box from "~app/atoms/Box";
import Button from "~app/atoms/Buttons";
import Input from "~app/atoms/Input";
import { useFullContext } from "~app/services/ContextProvider";
import { ActionTypes } from "~app/services/fetchProducts/types";

const SearchBar: FC = () => {
  const [, dispatch] = useFullContext();

  return (
    <Form
      onSubmit={values => {
        dispatch({ type: ActionTypes.FETCH, payload: values.search.trim() });
      }}
    >
      {({ handleSubmit }) => {
        return (
          <form onSubmit={handleSubmit} style={{ width: "100%" }}>
            <Box margin="auto" maxWidth="739px" width="100%">
              <Field
                name="search"
                validate={value => (value && value.trim() ? undefined : true)}
              >
                {({ input }) => {
                  return (
                    <Input
                      name={input.name}
                      onChange={input.onChange}
                      placeholder="Search"
                      value={input.value}
                    />
                  );
                }}
              </Field>
              <Box padding="0 0 0 10px">
                <Button type="submit">Search</Button>
              </Box>
            </Box>
          </form>
        );
      }}
    </Form>
  );
};

export default SearchBar;
