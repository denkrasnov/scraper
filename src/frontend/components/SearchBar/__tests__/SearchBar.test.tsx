import React from "react";
import { mount } from "enzyme";

import * as AppContext from "~app/services/ContextProvider";
import { INITIAL_STATE } from "~app/services/fetchProducts/constants";
import SearchBar from "..";

describe("SearchBar", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render", () => {
    jest
      .spyOn(AppContext, "useFullContext")
      .mockImplementation(() => [INITIAL_STATE, () => {}]);

    const component = mount(<SearchBar />);

    expect(component).toMatchSnapshot();
  });

  it("should submit the Form", () => {
    const dispatchMock = jest.fn();
    jest
      .spyOn(AppContext, "useFullContext")
      .mockImplementation(() => [INITIAL_STATE, dispatchMock]);

    const component = mount(<SearchBar />);

    const search = component.find("input[name='search']");
    ((search.instance() as unknown) as HTMLInputElement).value = "query";
    search.simulate("change", search);

    expect(component).toMatchSnapshot();

    component.find("form").simulate("submit");

    expect(dispatchMock).toHaveBeenCalledTimes(1);
  });
});
