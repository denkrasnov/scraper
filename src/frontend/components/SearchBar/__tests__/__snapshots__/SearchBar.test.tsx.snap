// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render 1`] = `
<SearchBar>
  <ReactFinalForm
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Box
        margin="auto"
        maxWidth="739px"
        width="100%"
      >
        <div
          className="Box"
          style={
            Object {
              "margin": "auto",
              "maxWidth": "739px",
              "width": "100%",
            }
          }
        >
          <Field
            name="search"
            validate={[Function]}
          >
            <Input
              name="search"
              onChange={[Function]}
              placeholder="Search"
              value=""
            >
              <input
                className="Input"
                name="search"
                onChange={[Function]}
                placeholder="Search"
                type="text"
                value=""
              />
            </Input>
          </Field>
          <Box
            padding="0 0 0 10px"
          >
            <div
              className="Box"
              style={
                Object {
                  "padding": "0 0 0 10px",
                }
              }
            >
              <Button
                type="submit"
              >
                <button
                  className="Button"
                  style={Object {}}
                  type="submit"
                >
                  Search
                </button>
              </Button>
            </div>
          </Box>
        </div>
      </Box>
    </form>
  </ReactFinalForm>
</SearchBar>
`;

exports[`SearchBar should submit the Form 1`] = `
<SearchBar>
  <ReactFinalForm
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Box
        margin="auto"
        maxWidth="739px"
        width="100%"
      >
        <div
          className="Box"
          style={
            Object {
              "margin": "auto",
              "maxWidth": "739px",
              "width": "100%",
            }
          }
        >
          <Field
            name="search"
            validate={[Function]}
          >
            <Input
              name="search"
              onChange={[Function]}
              placeholder="Search"
              value="query"
            >
              <input
                className="Input"
                name="search"
                onChange={[Function]}
                placeholder="Search"
                type="text"
                value="query"
              />
            </Input>
          </Field>
          <Box
            padding="0 0 0 10px"
          >
            <div
              className="Box"
              style={
                Object {
                  "padding": "0 0 0 10px",
                }
              }
            >
              <Button
                type="submit"
              >
                <button
                  className="Button"
                  style={Object {}}
                  type="submit"
                >
                  Search
                </button>
              </Button>
            </div>
          </Box>
        </div>
      </Box>
    </form>
  </ReactFinalForm>
</SearchBar>
`;
